### ISSUE TITLE
Equivalent mode identification during interpolation

### ISSUE THREAD

**Issue by g5t** [g5t/brille#25]
Following #22 correctly identifying equivalent modes is imperative for interpolation to give reasonable results. To this point efforts have focused on sorting the entire grid before interpolating under the assumption that
- a single grid-spanning permutation is possible
- sorting just the vertices needed for each interpolation point would be expensive

The single fully permuted grid may or may not be possible. Determining this with certainty is non-trivial; but what can be said is that the current sorting implementation does not produce a single-permutation grid.

What has not been investigated is whether the need for a single permutation can be avoided by determining the permutation within the interpolation step once the necessary vertices are known.
Doing so *may* be fast enough to not warrant optimization, but such an optimization route is clear:
1. (non-optimized) determine the permutation(s) needed to sort N<sub>modes</sub> on the N<sub>vertex</sub>-1 relative to the first needed vertex
2. sort the N<sub>vertex</sub> vertices (and the per-vertex weights) by vertex index, and cache the N<sub>vertex</sub>-1 length-N<sub>modes</sub> permutations using a thread-safe memoizer
2. give each thread an independent size-limited cache, using last-accessed time to decide which values to remove if the cache is full.

The initial approach (0<sup>th</sup> improvement) has the advantage of no significant memory footprint but if multiple Q points use the same set of vertices the same permutation will be determined multiple times.

The 1<sup>st</sup> improvement would ensure each unique set of permutations is determined only once, which is only an optimization if the cache lookup-time is less than the permutation determination time. The memory used by such a cache *could* be quite large since the interpolation uses between 1 and 8 vertices (or 1 and 4 if only tetrahedra are used), but thankfully is not as large as sum( {N<sub>vertex</sub>|i}, i=2,...,8) [where {N<sub>vertex</sub>|i} is the binomial coefficient, and we can always skip i=1 as it does not need a permutation] since it is restricted to combinations of vertices that are connected in the grid.

The 2<sup>nd</sup> improvement could help with large cache memory requirements and would work under the assumption that each thread works on Q points which are close to one another within the irreducible Brillouin zone grid.

If on-the-fly sorting works it might be possible to remove the attempt to construct a single-grid permutation.

**Comment by g5t** [comment]
The non-optimized on-the-fly equivalent mode identification has been implemented by #27

**Comment by g5t** [comment]
Now optimized following the ideas contained in list item 2 above. If further optimization is necessary it will require a new issue.
